Feature: HTTP following redirects

luismfonseca opened the pull request:
Fixes #1366 .

- [x] This PR is compliant with the [contributing guidelines](https://github.com/nodemcu/nodemcu-firmware/blob/dev/CONTRIBUTING.md) (if not, please describe why).
- [x] I have thoroughly tested my contribution.
- [x] The code changes are reflected in the documentation at `docs/en/*`.

Maybe the HTTP module should be configurable globally. As in `http.config(config_data)`, where you could set the user agent and wether or not to follow redirections.

Committers supporting this PR: leave blank

jmattsson commented:
Looks pretty good, but it would be nice to have some safeguard against infinite redirects. Do you have a strategy in mind?

luismfonseca commented:
Well, since the module doesn't handle the state per-connection, the only option really is to add a global static variable to count the number of redirects and give up after some reasonable value (i.e. Chrome does 21).

This module should be using a state per request so it could support concurrent requests. Then it would only be another field counting the redirects followed thus far.

I can do the first option if you want.

marcelstoer commented on app/http/httpclient.c:
```diff
@@ -307,26 +307,56 @@ static void ICACHE_FLASH_ATTR http_disconnect_callback( void * arg )
-					http_chunked_decode( body, chunked_decode_buffer );
-					os_memcpy( body, chunked_decode_buffer, body_size );
+
+				if ( locationOffset != NULL && http_status >= 300 && http_status <= 307 ) {
```
What about 308 i.e. RFC 7538?

luismfonseca replied:
You're right, I was only looking at the RFC 2616. I need to get with the times.

marcelstoer commented:
> Maybe the HTTP module should be configurable globally. As in `http.config(config_data)`...

Yes, that'd be useful indeed.

> I can do the first option if you want.

Go for it.

luismfonseca commented:
What I said before was true, the module now indeed has a session state per connection, so I just added a new variable there to count the number of redirects being followed.

I was also assuming the redirect path was a full url but it can be a relative path, so I implemented that as well.

Here are some sample requests to test all redirect types:
```
http.get("http://luisfonseca.xyz/redirect/infinite.php", nil, function(d,c) print(d,c) end, function(d,c) print(d,c) end) 
http.get("http://luisfonseca.xyz/redirect/absolute_path.php", nil, function(d,c) print(d,c) end, function(d,c) print(d,c) end) 
http.get("http://luisfonseca.xyz/redirect/relative_full_path.php", nil, function(d,c) print(d,c) end, function(d,c) print(d,c) end) 
http.get("http://luisfonseca.xyz/redirect/relative_same_folder.php", nil, function(d,c) print(d,c) end, function(d,c) print(d,c) end) 
```

marcelstoer commented on docs/en/modules/http.md:
```diff
@@ -13,6 +13,8 @@ Each request method takes a callback which is invoked when the response has been
 
 For each operation it is possible to provide custom HTTP headers or override standard headers. By default the `Host` header is deduced from the URL and `User-Agent` is `ESP8266`. Note, however, that the `Connection` header *can not* be overridden! It is always set to `close`.
 
+Redirects are followed automatically (note: the redirects will be followed up to a limit of 20 requests).
```
HTTP redirects (HTTP status 300-308) are followed automatically up to a limit of 20 to avoid the dreaded redirect loops.

marcelstoer commented:
Really nice, almost good to go.
